File: sdks/ts/packages/golem-ts-sdk/src/webhook.ts
// Copyright 2024-2025 Golem Cloud
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { createWebhook as createWebhookHost, PromiseId } from 'golem:agent/host@1.5.0';
import { createPromise, getPromise, GetPromiseResult } from 'golem:api/host@1.5.0';
import { awaitPromise } from './host/hostapi';

export function createWebhook(): WebhookHandler {
  const promiseId: PromiseId = createPromise();

  const webhookUrl = createWebhookHost(promiseId);

  return new WebhookHandler(webhookUrl, promiseId);
}

export class WebhookHandler implements PromiseLike<WebhookRequestPayload> {
  private readonly url: string;
  private readonly promiseId: PromiseId;

  constructor(url: string, promiseId: PromiseId) {
    this.url = url;
    this.promiseId = promiseId;
  }

  public getUrl(): string {
    return this.url;
  }

  private async wait(): Promise<WebhookRequestPayload> {
    const bytes = await awaitPromise(this.promiseId);

    return new WebhookRequestPayload(bytes);
  }

  then<TResult1 = WebhookRequestPayload, TResult2 = never>(
    onfulfilled?: ((value: WebhookRequestPayload) => TResult1 | PromiseLike<TResult1>) | null,
    onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | null,
  ): Promise<TResult1 | TResult2> {
    return this.wait().then(onfulfilled, onrejected);
  }
}

export class WebhookRequestPayload {
  private readonly payload: Uint8Array;

  constructor(payload: Uint8Array) {
    this.payload = payload;
  }

  public json<T>(): T {
    try {
      const text = new TextDecoder().decode(this.payload);
      return JSON.parse(text) as T;
    } catch (e) {
      throw new Error(`Invalid input: ${String(e)}`);
    }
  }

  public bytes(): Uint8Array {
    return this.payload;
  }
}

File: sdks/rust/golem-rust/src/agentic/webhook.rs
// Copyright 2024-2025 Golem Cloud
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::agentic::await_pollable;
use crate::bindings::golem::api::host::GetPromiseResult;
use crate::golem_agentic::golem::agent::host::AgentWebhook;
use crate::{get_promise, PromiseId};
use std::future::Future;
use std::future::IntoFuture;
use std::pin::Pin;
use std::str::FromStr;

pub fn create_webhook() -> WebhookHandler {
    let agent_webhook = crate::golem_agentic::golem::agent::host::create_webhook();

    let url = agent_webhook.get_callback_url();

    WebhookHandler::new(url, agent_webhook)
}

pub struct WebhookHandler {
    url: String,
    inner: AgentWebhook,
}

impl WebhookHandler {
    fn new(url: String, inner: AgentWebhook) -> WebhookHandler {
        WebhookHandler { url, inner }
    }
    async fn wait(self) -> WebhookRequestPayload {
        let promise_id = self.url.trim_end_matches('/').rsplit('/').next();

        match promise_id {
            Some(promise_id) => {
                let promise_id = PromiseId::from_str(promise_id).expect(&format!(
                    "Internal Error: Invalid webhook URL: {}",
                    self.url
                ));

                let promise_result: GetPromiseResult = get_promise(&promise_id);

                let pollable = self.inner.subscribe();

                await_pollable(pollable).await;

                let bytes = promise_result.get().unwrap();

                WebhookRequestPayload { payload: bytes }
            }

            None => {
                panic!("Internal Error: Invalid webhook URL: {}", self.url);
            }
        }
    }

    pub fn url(&self) -> &str {
        &self.url
    }
}

impl IntoFuture for WebhookHandler {
    type Output = WebhookRequestPayload;
    type IntoFuture = Pin<Box<dyn Future<Output = Self::Output>>>;

    fn into_future(self) -> Self::IntoFuture {
        Box::pin(async move { self.wait().await })
    }
}

pub struct WebhookRequestPayload {
    payload: Vec<u8>,
}

impl WebhookRequestPayload {
    pub fn json<T: serde::de::DeserializeOwned>(&self) -> Result<T, String> {
        serde_json::from_slice(&self.payload).map_err(|e| format!("Invalid input: {}", e))
    }
}
